pulser/simulation/simulation.py

```diff
@@ -325,55 +516,98 @@ def get_hamiltonian(self, time: float) -> qutip.Qobj:
-            progress_bar: Optional[bool] = None,
+    def run(self, progress_bar: Optional[bool] = None,
             **options: qutip.solver.Options) -> SimulationResults:
         """Simulates the sequence using QuTiP's solvers.
 
-        Keyword Args:
-            initial_state (array): The initial quantum state of the
-                evolution. Will be transformed into a ``qutip.Qobj`` instance.
-            progress_bar (bool): If True, the progress bar of QuTiP's
-                ``qutip.sesolve()`` will be shown.
-
-        Other Parameters:
-            options: Additional simulation settings. These correspond to the
-                keyword arguments of ``qutip.solver.Options`` for
-                the ``qutip.sesolve()`` method.
+        Will return NoisyResults if it detects any noise in the SimConfig.
+        Otherwise will return CleanResults.
 
-        Returns:
-            SimulationResults: Object containing the time evolution results.
+        Keyword Args:
+            progress_bar (bool): If True, the progress bar of QuTiP's solver
+                will be shown.
+            options (qutip.solver.Options): If specified, will override
+                SimConfig solver_options.
         """
-        if initial_state is not None:
-            if isinstance(initial_state, qutip.Qobj):
-                if initial_state.shape != (self.dim**self._size, 1):
-                    raise ValueError("Incompatible shape of initial_state")
-                self._initial_state = initial_state
+        solv_ops = (qutip.Options(max_step=5, **options) if options
+                    else self.config.solver_options)
+
+        def _assign_meas_basis() -> str:
+            if hasattr(self._seq, '_measurement'):
+                return cast(str, self._seq._measurement)
             else:
-                if initial_state.shape != (self.dim**self._size,):
-                    raise ValueError("Incompatible shape of initial_state")
-                self._initial_state = qutip.Qobj(initial_state)
-        else:
-            # by default, initial state is "ground" state of g-r basis.
-            all_ground = [self.basis['g'] for _ in range(self._size)]
-            self._initial_state = qutip.tensor(all_ground)
-
-        result = qutip.sesolve(self._hamiltonian,
-                               self._initial_state,
-                               self.eval_times,
-                               progress_bar=progress_bar,
-                               options=qutip.Options(max_step=5,
-                                                     **options)
-                               )
-        meas_basis: Optional[str]
-        if hasattr(self._seq, '_measurement'):
-            meas_basis = cast(str, self._seq._measurement)
+                if self.basis_name in {'digital', 'all'}:
+                    return 'digital'
+                else:
+                    return 'ground-rydberg'
```

Comment:
Can't you just define `meas_basis` here and use it for both the clean and noisy cases? It seems that the only point of this function is to give the same result to one of two variables: `measurement_basis` or `meas_basis`. What am I missing here?

Reply:
I think it should even be called right at initialisation. What do you think ?

Reply:
Yes, that would work.